### ISSUE TITLE
Add the ability to plot striped lines

### ISSUE THREAD

**Issue by odysseus9672** [matplotlib/matplotlib#18351]
### Problem
Sometimes plots (especially histograms with narrow bins) have small features that are obscured or missed by ordinary dashed line plots. 

### Proposed Solution
Add the ability to specify an `offcolor=(0.0, 0.0, 0.0, 0.0)` kwarg that fills in the blanks when a dashed line style, or dash pattern, is used. In principle, even more colors could be specified, but a simple on-color/off-color would be sufficient and require the fewest changes to the API.
This can already be done by plotting multiple lines that have carefully matching dash patterns, but the issue is that the stripes aren't represented in plot legends.

### Additional context and prior art
Right now I do this with this function (where the off-color is set to half the opacity of the `oncolor`):
```
def plotstripedline(ax, xvals, yvals, oncolor, dashpattern, lname):
    if len(oncolor) != 4:
        raise ValueError(
            "oncolor must have four floating point elements 0 to 1, rgba.")
    linestyle = (0, dashpattern)
    ax.plot(xvals, yvals, ls=linestyle, color=oncolor,
            label=lname)
    
    gapdp = list(dashpattern[1:])
    gapdp.append(dashpattern[0])
    gapstyle = (-dashpattern[0], gapdp)
    gapcol = [ x for x in oncolor ]
    gapcol[-1] *= 0.5
    ax.plot(xvals, yvals, ls=gapstyle, color=gapcol)
    
    return None
```
See, also, [this discussion on StackExchange](https://stackoverflow.com/questions/59130371/can-the-off-color-be-set-for-a-matplotlib-dashed-line).

**Comment by diegopetrola** [comment]
While this implementation is discussed, I believe you can make do with the following code:

```
import numpy as np
import matplotlib.pyplot as plt

x = np.arange(0, 2*np.pi, 0.2)
y = np.cos(x)

plt.plot(x, y, linestyle=(0, (2, 2)))
plt.plot(x, y, linestyle=(2, (2, 2)))

blue_line, = plt.plot(x, y, linestyle=(0, (2, 2)), lw=3, c=)
dashed_line, = plt.plot(x, y, linestyle=(2, (2, 2)), lw=3)
plt.legend([(blue_line, dashed_line)], ['data'])
```
![image](https://user-images.githubusercontent.com/3846018/91254403-b63f2d80-e738-11ea-9b66-25f4f66208d7.png)

I gave a quick look on the tutorials and it seems the legends in `matplotlib` are very powerful! If you wish, you can check more interesting examples [here](https://matplotlib.org/tutorials/intermediate/legend_guide.html).

**Comment by jklymak** [comment]
Someone would have to dig into linestyles to do this, and figure out a reasonable API for specifying the style.  Other stroke styles have been discussed in the past, so there may be a desire to make a "LineStyle" container class for specifying.  I'd say this is rather involved, but not an unreasonable request.

**Comment by WeatherGod** [comment]
Maybe it would make sense to borrow ideas from the filled two-color markers?

https://matplotlib.org/3.3.1/gallery/lines_bars_and_markers/marker_reference.html#marker-fill-styles

Or, alternatively, rethink how both of them could be done to have a more
conceptually consistent interface?

**Comment by tacaswell** [comment]
I agree with @jklymak 's assessment and @WeatherGod 

An immediate path would be to make a new artist that has as children the 2 lines in @diegopetrola 's comment.

**Comment by timhoffm** [comment]
> Maybe it would make sense to borrow ideas from the filled two-color markers?

API-wise an `alt_color` defaulting to transparent is enough.

**Comment by jklymak** [comment]
This sort of is similar to https://github.com/matplotlib/matplotlib/issues/12980, in particular https://github.com/matplotlib/matplotlib/issues/12980#issuecomment-448102430 which talks about pen and stroke styles.  I think it'd be a lot of work to do this, however.  

I guess `alt_color` is a possibility.  It would be less flexible, but certainly a more tractable problem.

**Comment by timhoffm** [comment]
> This sort of is similar to #12980, in particular #12980 (comment) which talks about pen and stroke styles. I think it'd be a lot of work to do this, however.

AFAICS #12980 is quite different. It has a single shaped pen, whereas here, you change between two differently colored "simple" pens. I don't think one of the problems can be expressed as the other.

But I agree that #12980 would be a lot of work, and `alt_color` is significantly simpler.

**Comment by WeatherGod** [comment]
for consistency, shouldn't the name be `coloralt`?

**Comment by timhoffm** [comment]
> for consistency, shouldn't the name be `coloralt`?

Maybe. Even though I find that quite unreadable. Alternatively, we could find something new like `gapcolor`.

**Comment by lucielavigne** [comment]
Hello, 
I am a student in software engineering and I work with a fellow student on a project. We would like to contribute on the implementation of this functionality

**Comment by timhoffm** [comment]
@lucielavigne You are welcome to!

### Technical remarks

I think the [new-artist-proposal](https://github.com/matplotlib/matplotlib/issues/18351#issuecomment-680994890) is not feasible. This would get us in all sorts of synchronization trouble:
- What if the user changes data -> would have to be synchronously propagated to both child artists.
- What should we do if the user changes linestyle from/to solid?

The correct solution is to dig into the low-level drawing and draw the spaces there as well.

We will need a discussion at some point on proper naming of the parameter (`gapcolor` / `coloralt` / `alt_color`), but that should not stop you to start writing that functionality. - Renaming is easy.

### Organizational remarks

**tl;dr: Please reflect if you've feel largely capable of the task. No offence meant. I'll assume you do, but want to make sure that we don't get into troubles down the road.**

We've had mixed experiences with student projects. Some resulted in valuable contributions, some ended up abandoned after a lot of back and forth. To prevent troubles down the road, I'd like to set expectations:

We welcome contributions from people with all experience levels. However, the difficulty of the task should match the experience. We'll happily give advice on specific aspects, but we're no teaching site (We don't have the capacity for requests "I'd like to work on this issue. Please walk me through step-by-step what I should do).
This issue is marked as medium-difficult. You should be able to find the relevant code locations and propose reasonable changes yourself. If you need help with some details, don't hesitate do ask.

CC @story645 as our community lead: Please check if wording and content matches our policy.

**Comment by story645** [comment]
Give or take - our code base is dense enough that I think it's okay to ask/confirm about relative code locations, but a lot of that thrashing will happen on the PR. @lucielavigne please don't take @timhoffm's response as dissuading you from contributing; if you're unsure about how much guidance you'll need, open a draft and we can take it from there. If you end up needing more help, we'll be at the PyData global sprints (which should be open to the public) https://global.pydata.org/talks/matplotlib

 Depending on how much support y'all need, we will be at the PyData global sprints (https://global.pydata.org/talks/matplotlib) which shouldn't require a ticket to get in.

**Comment by lucielavigne** [comment]
Hi, 
We are three students working on it, we have 6 weeks to work on it. We are willing to solve the issue.  
However the link you gave me is not working. 
We looked at the previous comments and the code, maybe you could give us some pointers to narrow our searchs. 
Thank you

**Comment by timhoffm** [comment]
I did not look closely at this, but I suppose you need to draw a second path with the inverted pattern at

https://github.com/matplotlib/matplotlib/blob/69286935b8cd122d2b7d529bc075b1fecdf021a1/lib/matplotlib/lines.py#L787.

**Comment by story645** [comment]
Sorry, the new link is https://global.pydata.org/talks/99, but  there aren't details posted yet so check back at https://gitter.im/matplotlib/matplotlib

**Comment by lucielavigne** [comment]
Hello, 
I submitted a pull request under add the ability to plot stripped lines.
